src/main/java/net/fabricmc/installer/Handler.java

```diff
@@ -92,63 +127,122 @@ public JPanel makePanel(InstallerGui installerGui) {
 			});
 		});
 
+		
 		Main.LOADER_META.onComplete(versions -> {
-			for (int i = 0; i < versions.size(); i++) {
-				String version = versions.get(i);
-				loaderVersionComboBox.addItem(version);
-			}
-			if (loaderVersionComboBox.getItemCount() > 0) loaderVersionComboBox.setSelectedIndex(0);
+			updateLoaderVersions();
 			statusLabel.setText(Utils.BUNDLE.getString("prompt.ready.install"));
 		});
 
 		return pane;
 	}
 
+	private static int compareVersions(String a, String b) {
+		if (a == null) return -1;
+		if (b == null) return 1;
+		if (a.equals(b)) return 0;
+		Pattern pattern = Pattern.compile("^\\d+(\\.\\d+)+");
+		Matcher matcherA = pattern.matcher(a);
+		Matcher matcherB = pattern.matcher(b);
+		if (!matcherA.find() || !matcherB.find()) {
+			throw new IllegalArgumentException("Invalid version");
+		}
+		String[] aParts =  matcherA.group(0).split("\\.");
+		String[] bParts =  matcherB.group(0).split("\\.");
+		for (int i = 0; i < aParts.length || i < bParts.length; i++) {
+			if (i >= aParts.length && !"0".equals(bParts[i])) return -1;
+			if (i >= bParts.length && !"0".equals(aParts[i])) return 1;
+			int cmp = Integer.compare(Integer.parseInt(aParts[i]), Integer.parseInt(bParts[i]));
+			if (cmp != 0) return cmp;
+		}
+		return 0;
+	}
+	
+	private void updateLoaderVersions()
+	{
+		loaderVersionComboBox.removeAllItems();
+		for (MetaHandler.GameVersion version : Main.LOADER_META.getVersions())
+		{
+			boolean guavaLoader = ((MetaHandler.LoaderVersion) version).getName().equals("fabric-loader-1.8.9");
+			String gameVersion = (String) gameVersionComboBox.getSelectedItem();
+			String loaderVersion = version.getVersion();
+			try {
+				// Versions < 1.12 ship Guava 17 so they need the fix
+				if (compareVersions(gameVersion, "1.12") < 0) {
+					if (guavaLoader) {
+						loaderVersionComboBox.addItem(loaderVersion);
+						loaderVersionComboBox.setSelectedItem(version.isStable() ? loaderVersion : 0);
+					}
+					continue;
+				}
+				// 1.12 needs a fix that's implemented since loader 0.11.3
+				if (compareVersions(gameVersion, "1.12.2") <= 0) {
+					if (compareVersions(loaderVersion, "0.11.3") >= 0) {
+						loaderVersionComboBox.addItem(loaderVersion);
+						loaderVersionComboBox.setSelectedItem(version.isStable() ? loaderVersion : 0);
+					}
+					continue;
+				}
```

Comment:
Is this really necessary since meta does the same checks? You'd somehow have to pass the `gameVersion` to `LOADER_META.getVersions()`